### ISSUE TITLE
Broadcast receiver won't receive any activity transition

### ISSUE THREAD

**Issue by loresuso** [googlecodelabs/activity_transitionapi-codelab#9]
I tested this application, but it seems it doesn't work. (Cloned directly from this repo)

**Comment by ManasMalla** [comment]
For me too...! Running on a Mi A2 and Mi A3

**Comment by loresuso** [comment]
I solved my problems by not using at all the transition fashion and using the periodic one. That worked for me. So instead of calling requestActivityTransitionUpdates() you need to call requestActivityUpdates() with a given interval of time.

**Comment by josh-bartz** [comment]
I've got the same problem. Galaxy s8 on Android 9 and Pixel 3 on Android 11. Tried @loresuso's workaround, and still nothing.

**Comment by stewartmcm** [comment]
same problem on a Pixel 2/Android 10. I also tried @loresuso 's modification, but no luck.

**Comment by loresuso** [comment]
Ok I think I didn't explain myself well. Here is what I intended: 

In main activity:
<code>
final int PERIOD = 1000; // in ms 
ActivityRecognitionClient client = new ActivityRecognitionClient(this);
Intent i = new Intent(this, MyIntentService.class);
PendingIntent pi = PendingIntent.getService(this, 1, i, PendingIntent.FLAG_UPDATE_CURRENT);
Task<Void> task = client.requestActivityUpdates(PERIOD, pi)
</code>

Now every second will arrive an Intent to the IntentService MyIntentService you may define. From it, you can extract the result of the activity recognition api by simply 
<code>
ActivityRecognitionResult result = ActivityRecognitionResult.extractResult(intent);
ArrayList<DetectedActivity> activities = result.getProbableActivities();
</code>

This should do the trick, let me know it!

**Comment by josh-bartz** [comment]
No dice. Tried with the broadCastReceiver and an IntentService. Thanks for trying, though.

**Comment by stewartmcm** [comment]
@loresuso agree that using an IntentService works for handling activity updates, but IntentService was deprecated with sdkVersion 30. Android docs recommend a BroadcastReceiver (like in the codelab) but looks like LocalBroadcastManager is deprecated too? (though docs don't specify at which version that will take effect).

I'll try filing a bug with issue tracker.

**Comment by loresuso** [comment]
I'm sorry it didn't work!  It's weird 'cause I tested my app in many devices, including also Android 10, and it worked perfectly. To replicate exactly my environment, I have to say that the initialization of activity recognition client is in a foreground service in my app. I don't know if that is a useful information, but I think they have recently introduced limitations on the execution of backgound jobs, so _maybe_ this could be relevant.

Yes I know IntentService was deprecated. They say it could be replaced with JobIntentService, but I could not figure out _how._ With the JobIntentService you have to enqueue the job, and you have to do it only when the intent is received. I miss the part on receiving the intent itself ..

Hope you can find a workaround!

**Comment by loresuso** [comment]
I'm sorry it didn't work. Hope you can find a workaround!

**Comment by emreakcan** [comment]
@stewartmcm @josh-bartz 
please try my pull request 
https://github.com/emreakcan/activity_transitionapi-codelab/tree/issue_9

**Comment by josh-bartz** [comment]
> @stewartmcm @josh-bartz
> please try my pull request
> https://github.com/googlecodelabs/activity_transitionapi-codelab/compare/master...emreakcan:issue_9?expand=1

Worked for me. Thanks, @emreakcan

**Comment by emreakcan** [comment]
@loresuso  why did you close that? it's not merged yet.

**Comment by loresuso** [comment]
@emreakcan I'm sorry, I'm new to github. I tested your code and it worked, so I thought I could just close it.

**Comment by codingjeremy** [comment]
Hi all,

Thanks for all the great feedback on this issue!

The `complete` module for this codelab is working again, so please take a look and reopen this if it isn't working for you. I also updated the codelab doc. Please note, it is still using `requestActivityTransitionUpdates()`.
 
Please test on a physical device as using the emulator to recreate activity transitions isn't working consistently for me right now and I verified with engineers that tests use physical devices. Also, you may notice the first time you run it that all the events won't start coming in right away (can talk 10 seconds to about a minute), so keep that in mind.

Why did it stop working?

I think it was a couple things. First, the LocalBroadcastManager stopped picking up events. This initially worked when the codelab was working in late 2019, but it stopped working more recently. LBM is being deprecated as some of you pointed out, so I just changed that out to a regular register/unregister.

I noticed the initial version created the BR in onCreate() and there is a chance with it being changed to null in onStop(), that it wouldn't be recreated. I fixed that. I also did some minor updates (cleaned up UI, triggered subscribing to activity recognition if you accept permissions, etc.).

Let me know if I missed anything, thanks!

**Comment by markwilcox** [comment]
Just for info if others are having trouble getting activity transitions to work, I based an implementation on this project and couldn't get it working on some devices. I then cloned this repo and installed the "complete" code lab project on the same devices - it doesn't work at all. The devices were a fairly recent Xiaomi Mi 9 SE and an old Samsung Galaxy S5 (Android 5 - had to reduce the minimum API to 21 and remove a setTextAppearance call from the LogFragment to make that work). I have tried walking around with each for a few minutes, and I've tried adding the other activity types and doing those activities - I don't get transitions for any of them - well I did get a STILL Enter once on the Xiaomi device.

I have a suspicion that the fancy machine learning models for Activity Transitions don't work with all device sensor readings. On the old Samsung I've seen the Activity Updates (rather than transitions) pop up and tell me with 90% confidence that the device is in a vehicle, having not moved on my desk for a few minutes.

As far as I can tell from the documentation, the latest implementation here is how it's intended to work (with LBM and IntentService deprecated, and the docs do explicitly say to use a BroadcastReceiver).
